lib/shelley/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs

```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+--
+-- Exposes a wallet-friendly interface to ledger types and functions.
+--
+module Cardano.Wallet.Shelley.Compatibility.Ledger
+    (
+      -- * Exported functions
+      computeMinimumAdaQuantity
+
+      -- * Conversion between wallet types and ledger types
+    , HasLedgerType (..)
+
+    ) where
+
+import Prelude
+
+import Cardano.Crypto.Hash
+    ( hashFromBytes, hashToBytes )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName (..), TokenPolicyId (..) )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity (..) )
+import Data.Function
+    ( (&) )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.Word
+    ( Word64 )
+import Fmt
+    ( pretty )
+import GHC.Stack
+    ( HasCallStack )
+import Ouroboros.Consensus.Shelley.Protocol.Crypto
+    ( StandardCrypto )
+
+import qualified Cardano.Ledger.Mary.Value as Ledger
+import qualified Cardano.Ledger.ShelleyMA.Rules.Utxo as Ledger
+import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
+import qualified Data.Map.Strict as Map
+import qualified Data.Map.Strict.NonEmptyMap as NonEmptyMap
+import qualified Shelley.Spec.Ledger.API as Ledger
+
+--------------------------------------------------------------------------------
+-- Exported functions
+--------------------------------------------------------------------------------
+
+-- | Uses the ledger specification to compute the minimum required ada quantity
+--   for a token bundle.
+--
+computeMinimumAdaQuantity
+    :: TokenBundle
+    -- ^ The token bundle to evaluate.
+    -> Coin
+    -- ^ The absolute minimum ada quantity specified by the protocol.
+    -> Coin
+    -- ^ The minimum ada quantity for the given token bundle.
+computeMinimumAdaQuantity bundle protocolMinimum =
+    fromLedger @Coin $
+        Ledger.scaledMinDeposit
+            (toLedger @TokenBundle bundle)
+            (toLedger @Coin protocolMinimum)
+
+--------------------------------------------------------------------------------
+-- Conversions between wallet types and ledger types
+--------------------------------------------------------------------------------
+
+class HasLedgerType wallet ledger | wallet -> ledger where
+    -- | Converts a value from a ledger type to the equivalent wallet type.
+    fromLedger
+        :: HasCallStack => ledger -> wallet
+    -- | Converts a value from a wallet type to the equivalent ledger type.
+    toLedger
+        :: HasCallStack => wallet -> ledger
```

Comment:
This deviates quite a bit from the approach already taken in `.../Shelley/Compatibility.hs`. May we stick with the same approach and have these conversions in `.../Shelley/Compatibility.hs` following the existing naming conventions? 

We don't need a type-class for this, plain functions are more readable and easily to pass around.gh There are already far enough constraints from integrating with cardano-api and the cardano-ledger-specs :+1:

Reply:
Hi Matthias

Thanks for your feedback. 👍🏻 

> plain functions are more readable and easily passed around.

I've added https://github.com/input-output-hk/cardano-wallet/pull/2461/commits/0d96de1ebe8a39a4f1117a9b7aa7796b858b429e, which exports plain functions for each of the conversions, as requested:

```hs
      -- * Conversions from wallet types to ledger specification types
    , toLedgerCoin
    , toLedgerTokenBundle
    , toLedgerTokenPolicyId
    , toLedgerTokenName
    , toLedgerTokenQuantity

      -- * Conversions from ledger specification types to wallet types
    , toWalletCoin
    , toWalletTokenBundle
    , toWalletTokenPolicyId
    , toWalletTokenName
    , toWalletTokenQuantity
```

> We don't need a type-class for this

There's actually no need for callers of this module to use any type classes. They can just use the plain functions, which are now exported (see above).

My intention here was to encode the fact that certain pairs of types support round-trip conversion, which in turn allows concise round-trip testing:

```hs
describe "Roundtrip conversions between wallet and ledger specification types" $ do
    ledgerRoundtrip $ Proxy @Coin
    ledgerRoundtrip $ Proxy @TokenBundle
    ledgerRoundtrip $ Proxy @TokenName
    ledgerRoundtrip $ Proxy @TokenPolicyId
    ledgerRoundtrip $ Proxy @TokenQuantity
```

The expected properties can now be documented in _just one place_, rather than duplicating these properties in the comments of every single conversion function:
```hs
-- | A pairing between a wallet type and its equivalent ledger type.
--
-- Instances of this class should satisfy the following laws:
--
-- >>> toLedger . toWallet == id
-- >>> toWallet . toLedger == id
--
class Convert wallet ledger | wallet -> ledger
```

> This deviates quite a bit from the approach already taken in .../Shelley/Compatibility.hs.
> May we stick with the same approach and have these conversions in .../Shelley/Compatibility.hs following the existing naming conventions?

My reason for putting this in a separate module is:

- **`Compatibility`** seems to be mainly concerned with the **node**, and the types and functions it exports. (Though there might be some other stuff in there too.)
- **`Compatibility.Ledger`** is concerned with the **ledger specification**, which has completely different types and encodings for certain things (such as token bundles).

These two layers seem distinct enough to warrant having separate modules, wouldn't you say?

In general, assuming concepts can be cleanly separated, I'm not a big fan of very large modules that mix different concepts together. In addition, the `Compatibility` module is already huge: it's nearly 1500 lines!

Reply:
Compatibility currently concerns cardano-api, ouroboros-network and cardano-ledger-specs. Basically, the cardano's ecosystem gravitating around cardano-node. I don't see any reason to put these in a separate module or to treat them differently than the rest. 

> I'm not a big fan of very large modules that mix different concepts together. In addition, the Compatibility module is already huge: it's nearly 1500 lines!

I know that. But if we want to split this module, we better do it consistently and with a more holistic approach; and not have _some_ functions in `Compatibility` and some others in `Compatibility.Ledger`. Now that just adds to the overall complexity. It also save us hardly any compilation time because both modules are needed around the same time to compile the final executable, and both comes already at the end of the compilation stack.